### ISSUE TITLE
Bundle decorator to run the check

### ISSUE THREAD

**Issue by tuukkamustonen** [alisaifee/flask-limiter#67]
I have disabled `auto_check` because I want to trigger rate-limit check at a specific point during request processing.

My routes are something like this:

```python
@app.route(...)
@auth.auth(...)
...
def route():
    ...
```

And I want limiter to run after auth check. So I added something like:

```python
class ExtLimiter(Limiter):

    def rate_limit(self):
        def wrapper(f):
            @functools.wraps(f)
            def wrapped(*args, **kwargs):
                self.check()
                return f(*args, **kwargs)
            return wrapped
        return wrapper
```

And now I:

```python
@app.route(...)
@auth.auth(...)
@limiter.rate_limit()
...
def route():
    ...
```

I think this simple decorator could be bundled in Flask-Limiter?

(I was kinda surprised when `@limiter.limit()` only registered the route and did not run the `check()` at all. So, added this decorator.)

**Comment by blurrcat** [comment]
Same here. I assumed the decorator would run the check too, and it failed to access `current_user`. 
Currently, I'm using this with `auto_check=False` as a workaround:
```py
def limit_and_check(
    limit_value, key_func=None, per_method=False, methods=None,
    error_message=None, exempt_when=None,
):

    def inner(func):
        func = limiter.limit(
            limit_value, key_func=key_func, per_method=per_method,
            methods=None, error_message=None, exempt_when=exempt_when,
        )(func)

        @wraps(func)
        def check(*args, **kwargs):
            limiter.check()
            return func(*args, **kwargs)

        return check

    return inner
```

**Comment by alisaifee** [comment]
Thanks for the suggestions, and since two of you had similar requirements - could I recommend that  one of you open a pull request? I'd be happy to work on it with you guys to get it merged.

**Comment by blurrcat** [comment]
What do you think about my workaround? It requires `auto_check=False` though, which is not the default.
Speaking of `auth_check`, since it's in a `before_request` hook, the check always runs before view decorators, like `login_required` from flask-login. This means `current_user` in the key function won't work if `auto_check` is left to True.

I think the ideal behavior would be:
1. if a view is decorated by `limit`, run check in the limit decorator instead of `before_request`.
2. otherwise, auto check with the default key function

Thoughts?

**Comment by danielstocks** [comment]
I'm also intersted in this, since i was to be able to run rate-limiting before stuff like data/schema validation and user auth.

I thought this could be done because in the docs i found this recipe:
https://flask-limiter.readthedocs.io/en/stable/#rate-limit-key-functions

```python
@route("/test")
@login_required
@limiter.limit("1 per day", key_func = lambda : current_user.username)
def test_route():
    return "42"
```

I assume this won't work though, because the limiter check is actually applied before `login_required` is run, am i correct? I had to apply the same patch as @tuukkamustonen @blurrcat to work around this issue. 

To my mind i was just confused by the `auto_check` option, and naturally thought that the limiter decorator would run in specified order in relation to the other view decorators. To change this would literally break all existing implementations so i guess that's not a good idea either.

**Comment by danielstocks** [comment]
> I think the ideal behavior would be:
> 1. if a view is decorated by limit, run check in the limit decorator instead of before_request.
> 2. otherwise, auto check with the default key function
> Thoughts?

^ I actually think this makes a lot more sense than the current implementation

**Comment by alisaifee** [comment]
Didn't actually mean to close the issue :) 
I was wondering if you guys could take the code in master for a test run with your use cases to see if things are as expected.

**Comment by danielstocks** [comment]
Awesome! Thanks @alisaifee. I will try it out and let you if it works!

**Comment by alisaifee** [comment]
I've made a pre-release .. release on pypi as well if that makes it easier to test out:
` pip install flask-limiter==1.0.0rc1`

**Comment by ncrocfer** [comment]
Great, it works fine for me ! Thanks @alisaifee ;)

**Comment by tuukkamustonen** [comment]
To answer myself:

It looks like `auto_check=True` is required for decorator to do anything (https://github.com/alisaifee/flask-limiter/blob/master/flask_limiter/extension.py#L541).

And yeah, we don't need to "register" the endpoints anymore, so following is sufficient:

```python
bp = Blueprint('foobar', __name__)

@bp.route(...)
@limiter.limit("1/s", key_func=get_limiter_key)
def foo():
    return "bar"
```

Or to re-use same counter over endpoints:

```python
bp = Blueprint('foobar', __name__)

limit = limiter.shared_limit('1/s', 'scope', key_func=get_limiter_key)

```python
@bp.route(...)
@limit
def foo():
    return "bar"

@bp.route(...)
@limit
def bar():
    return "foo"
```

I thought `before_request` would be called before hitting the decorators, so that if I had `auto_check=True`, limiting would happen in `before_request` and not in @decorator (not something we want), and looks like that really happens (`__check_request_limit()` gets called twice per request). But when called through `before_request`, `__check_request_limit()` gets emptyt `all_limits` list (https://github.com/alisaifee/flask-limiter/blob/master/flask_limiter/extension.py#L462), so nothing happens. Only when decorator triggers `__check_request_limit()` the second time, `all_limits` get data and rate-limiting gets applied.

Now, I'm not saying that it's bugging or that it's wrong, but doesn't it feel a bit wasteful?

**Comment by sbnetops** [comment]
I am facing a similar issue. I don't want rate limit checks in `before_request`, but only after some custom decorators have added stuff to the request context. In order to do that, I had to change `auto_check=False`. But apparently this line https://github.com/alisaifee/flask-limiter/blob/c0ae4d944bf233dccd68207c4de9e2c242cb055f/flask_limiter/extension.py#L564 disables rate limit checking completely when `auto_check=False`, even if we explicitly marked the route to be rate-limited with the decorator `limiter.limit()`.

The workaround I have found is to explicitly set `_auto_check` after the object is created and there is not chance to register the rate-limit check in the `before_request`, but it won't stop decorated endpoints from being checked.

```python
limiter = Limiter(
    web,
    key_func=ratelimit.rate_limit_key,
    default_limits=[ratelimit.rate_limit_by_tier],
    auto_check=False,
)
limiter._auto_check = True
```

I believe the `auto_check` should only disable registering to `before_request` and not all decorated routes.
